RoT/Protocol/RFCs/0001-Transport_Agnosticism.md

```diff
@@ -0,0 +1,149 @@
+  as an example admissible transport layer for I2C connections. The same should
+  be done for Section 8.1, which describes a legacy protocol built on top of
+  SMBus.
+
+- The first few sections of Chapter 6 should be modified to not reference MTCP,
+  and instead refer to the abstract Cerberus header (request bit, type, length)
+  derived from the transport-specific header.
+
+- The MCTP-specific errors in Section 6.5 should be either removed, or made
+  transport agnostic. For example, the packet/message length overflow errors
+  should be handled at the transport protocol layer: in the case that a device
+  sends a too-large packet, Cerberus should not be notified at all.
+
+- Mentions of MCTP in Section 6.7 should be removed. The maximum message/packet
+  size fields should be removed, since the transport layer should negotiate
+  these while establishing a session, and pass the maximum message size field
+  along to Cerberus.
+
+- The same should be done for Section 6.13.
+
+- The same should also be done for Section 6.18. Since the response makes
+  reference to payload sizes, we recommend that, instead of replying with
+  multiple messages of maximum size, we add a "continue" bit to the Get Log
+  response. When this bit is set, the client must send another request, with
+  the offset field increased by the length of the Get Log log contents field.
+
+- The same applies to Sections 6.20, 6.23, and 6.31.
+
+# Implementation Guidance
+
+Manticore currently implements the above proposal via the following procedure:
+1. Recieve a vtable from the library user that provides a function for blocking
+   until a request arrives from somewhere.
+
+2. When that function returns, Manticore's machinery calls a different function
+   in the vtable for parsing the parts of the transport-specific header it needs
+   to select a message parser and handler (request bit and type byte).
+
+3. Manticore calls into the appropriate parser, which calls a third function in
+   the vtable that provides a `read(2)` interface over the message payload to
+   parse the message.
+
+4. Pass the parsed message into the appropriate request handler, which constructs
+   a response.
+
+5. Manticore then calls a fourth function in the vtable, passing along the type
+   of the response, that instructs the transport to prepare for a response. Note
+   that the original request contained addressing information that the vtable
+   tracks and re-uses for addressing the reply.
+
+6. The transport encodes a response header; Manticore's serializer calls into a
+   `write(2)` interface to encode the response. The transport may packetize
+   the response as its internal buffer fills up.
+
+7. Manticore calls the final, sixth function to flush the response, sending the
+   final packet.
+
+As the author understands it, this is somewhat close to how the Microsoft
+implementation handles messages, although it calls the MCTP library directly
+rather than virtually.
+
```

Comment:
Our implementation is the following:
1. cmd_channel.receive_packet is called to get a packet of data from the transport (we are agnostic to the physical layer).
2. When a packet is received, it is passed to the the MCTP stack through mctp_interface_process_packet.  As you say, this is a direct call instead of a virtual one.
3. The MCTP stack will parse the packet and work to reconstruct the complete message.
4. If the current packet completes an MCTP message, the message (stripped of MCTP packet headers but not the MCTP vendor-specific command header) is sent for command processing through cmd_interface.process_request.
5. If the message was an MCTP control message instead of a Cerberus one, it would be handled in the MCTP stack.  MCTP errors are also directly handled by the MCTP stack.
6. The response from command processing is packetized by the MCTP stack and returned to the cmd_channel for transmission.
7. cmd_channel will loop through the list of packets and send them using cmd_channel.send_packet.  There is no understanding of MCTP structure in the cmd_channel.

This Microsoft implementation seems to align with the suggested changes here, with a couple of minor modifications:
- Make the transport protocol stack a function pointer call instead of a direct call into MCTP handling.
- Add vender-specific commands to the MCTP stack to support message/packet size and timeout detection.